### PR TITLE
feat: add maxProperties declarative validation tag

### LINKED ISSUE #138112: Declarative Validation: Add `+k8s:maxProperties` declarative validation tag
# Add `+k8s:maxProperties` declarative validation tag

## Summary

Add a new `+k8s:maxProperties=<int>` declarative validation (DV) tag that enforces a maximum number of keys in a map field. This corresponds to the OpenAPI/JSON Schema `maxProperties` keyword.

## Motivation

OpenAPI/JSON Schema supports `maxProperties` as a standard validation keyword. DV currently has no equivalent. Several Kubernetes API map fields have upper bounds on the number of entries for resource/performance reasons, currently enforced with handwritten validation.

## Kubernetes API Example

**`StorageClass.Parameters`** in `pkg/apis/storage/validation/validation.go`:

```go
// Handwritten validation today:
const maxProvisionerParameterLen = 512

if len(params) > maxProvisionerParameterLen {
    allErrs = append(allErrs, field.TooLong(fldPath, "" /*unused*/, maxProvisionerParameterLen))
    return allErrs
}
```

With the proposed tag:

```go
type StorageClass struct {
    // +k8s:maxProperties=512
    Parameters map[string]string
}
```

## OpenAPI Semantics

- **OpenAPI keyword:** `maxProperties`
- **Go type in JSONSchemaProps:** `*int64`
- **Spec behavior:** The object (map) must have at most the specified number of properties/keys.

## DV Design Considerations

- **Map types only:** Same as `+k8s:minProperties` — only meaningful for `map` types in Go, since struct fields are fixed at compile time. Type check should restrict to `types.Map` kind.
- **Relationship to `+k8s:maxItems`:** `+k8s:maxItems` validates slice/array length and uses the **ShortCircuit flag** to stop further validation if the limit is exceeded (prevents DoS via oversized inputs). `+k8s:maxProperties` should similarly consider short-circuiting — an oversized map should skip per-key/per-value validation.
- **Value parsing:** Should use `util.ParseInt()` like other limit tags. Value must be non-negative.

## Implementation Notes

- Should be implemented in `validators/limits.go` alongside existing min/max tags
- Consider using `ShortCircuit` flag like `+k8s:maxItems` to avoid validating entries in oversized maps
- Should generate a call to a validation function in `k8s.io/apimachinery/pkg/api/validate`
- Validator file: `staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go`

/sig api-machinery
/triage accepted
/area api-validation

### PR DESCRIPTION
**Pull request by aaron-prindle**
#### What type of PR is this?

/kind feature
/area api-validation

#### What this PR does / why we need it:

Adds a new declarative validation tag, `k8s:maxProperties`, that ensures that maps do not contain more than X properties.

This PR cherry-picks the changes from `jpbetz/validation-gen#211` (original author @everettraven, thanks!) into the k/k repository, adding support for the `+k8s:maxProperties` tag in `validation-gen` and the corresponding validator in `apimachinery`. Additional changes were necessary though in upstreaming the PR, see the reviewer notes below.

#### Which issue(s) this PR is related to:

Fixes #138112

#### Special notes for your reviewer:
`+k8s:maxProperties` will be added as a `StabilityLevelBeta` tag (vs *Alpha) as it is based off of the Open API tag of the same name and the tag API is well defined (not subject to change).

Made a few adjustments to the original feature branch PR changes to satisfy the k/k repository's linter and existing code patterns:

1.  In `staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/tags/maxproperties/doc.go`, used `// +k8s:validation-gen-nolint` to satisfy the linter (same practice other tests use).
2.  In `staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go`, used `util.ParseInt` instead of `strconv.Atoi` to match existing patterns, and `sets.List` instead of `UnsortedList` due to the version of the `sets` package in use.
3.  In `staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go`, rewrote `TestMaxProperties` to use the standard `field.ErrorMatcher` pattern instead of the custom helper used in the original PR.
4.  In `staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go`, fixed the type-checking logic for map keys to correctly support typedefs to string (like `type StringKey string`) and avoid potential nil pointer dereference panics for non-map types during generation.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
N/A
```

/triage accepted

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go [hunk 1]
```diff
@@ -210,6 +210,57 @@ func TestMaxItems(t *testing.T) {
...
+		name:       "2 properties, max 1",
+		properties: 2,
+		max:        1,
+		wantErrs: field.ErrorList{
+			field.TooMany(field.NewPath("fldpath"), 2, 1).WithOrigin("maxProperties"),
+		},
+	}, {
+		name:       "0 properties, max -1",
+		properties: 0,
+		max:        -1,
```
Comment (jpbetz) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:244]: I don't believe json schema allows maxProperties to be `-1`.  Testing that the limit check still works for this is fine, but a comment explaining seems justified.

File: staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go [hunk 2]
```diff
@@ -262,6 +264,65 @@ func (mitv maxItemsTagValidator) Docs() TagDoc {
...
+	if intVal < 0 {
+		return result, fmt.Errorf("must be greater than or equal to zero")
+	}
+	// Note: maxProperties short-circuits other validations.
+	result.AddFunction(Function(maxPropertiesTagName, ShortCircuit, maxPropertiesValidator, intVal))
+	return result, nil
+}
+
+func (mptv maxPropertiesTagValidator) Docs() TagDoc {
+	return TagDoc{
```
Comment (jpbetz) [inline staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go:312]: I'd prefer we consistently either use golang terminology such as "map" and "element", or we consistently use JSON schema terminology such as "object" and "properties".

Or, we could explain it explicitly in both terms, e.g.:

"maxProperties provides a limit on properties of an object as defined by JSON schema.  In Kubernetes is may only be used to constrain the number of elements on a field defined as a golang map.".

File: staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go [hunk 3]
```diff
@@ -210,6 +210,57 @@ func TestMaxItems(t *testing.T) {
 	}
 }
 
+func TestMaxProperties(t *testing.T) {
```
Comment (jpbetz) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: For completeness, let's also test the upper bound (max int).

File: staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go (hunk 1 above)
Comment (aaron-prindle) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:257]: Added a comment explaining the test despite JSON Schema not allowing this:
```
		// Note: While JSON Schema does not allow negative values for maxProperties,
		// we test that the validator handles it safely if it ever occurs at runtime.
```

We have this check on the input:

```go
    if intVal < 0 {
        return result, fmt.Errorf("must be greater than or equal to zero")
    }
```
so this shouldn't be possible in practice.

File: staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go (hunk 2 above)
Comment (aaron-prindle) [reply staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go:315]: Updated the `Description:` to use both terms explicitly with the text suggested: 

> "maxProperties provides a limit on properties of an object as defined by JSON schema. In Kubernetes it may only be used to constrain the number of elements on a field defined as a golang map."

File: staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go (hunk 3 above)
Comment (aaron-prindle) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: Done:
```go
	}, {
		name:       "0 properties, max MaxInt",
		properties: 0,
		max:        math.MaxInt,
	}, {
		name:       "1 property, max MaxInt",
		properties: 1,
		max:        math.MaxInt,
	}, {
```

Comment (jpbetz) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: That doesn't actually test the edge fully. I was expecting cases like:

```go
{
		properties: math.MaxInt-1,
		max:        math.MaxInt,
}, {
		properties: math.MaxInt,
		max:        math.MaxInt,
}, {
		properties: math.MaxInt,
		max:        math.MaxInt-1,
		wantErrs ...
}
``

Comment (aaron-prindle) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: Yes sorry, good point I should have explained the friction I hit more and how I ended up with those test cases as you have a good point they don't test the most important edge cases w.r.t. `math.MaxInt`directly.  When I attempted to create a map with `math.MaxInt` (as well as `math.MaxInt - 1`) properties in Go, the way the test is currently constructed it creates actual maps and requires allocating that many elements.  For these cases this causes a timeout panic initializing the math.MaxInt size map (2^(63-1))

Example:
```go
              {
		name:       "math.MaxInt property, max MaxInt-1",
		properties: math.MaxInt,
		max:        math.MaxInt - 1,
		wantErrs: field.ErrorList{
			field.TooMany(field.NewPath("fldpath"), math.MaxInt, math.MaxInt-1).WithOrigin("maxProperties"),
		},
              }
```
logs: https://gist.github.com/aaron-prindle/4ec33aadfbd60354fa7cd4b0b0ed39e0
log snippet w/ the timeout:
```
panic: test timed out after 30s
	running tests:
		TestMaxProperties (30s)
		TestMaxProperties/math.MaxInt_property,_max_MaxInt-1 (30s)
```

As Go's `len()` function only works on actual built-in maps and can't be directly mocked IIUC, it is non-trivial to practically create a map w/ math.MaxInt size to trigger the failure condition for `max: math.MaxInt - 1`.

The tests cases there w/ `properties: [0|1]` + `max: math.MaxInt` were an attempt to use math.MaxInt w/ empty and set maps to hit some testing surface but understandably this doesn't hit the edge as noted. 

I believe the alternative options to make the test work here would be.  Most have a drawback or are overly complex vs keeping the valdiator itself as is.  For modifying the test only we can use `unsafe` to spoof a  large map but using unsafe :

1. Update `MaxProperties` to accept `any` and use reflection to get the length.
    - **Drawback w/ this approach**: This avoids any changes to the `validation-gen` framework (map can be passed as `any`). However there is performance penalty at runtime as it relies on reflection to determine the map's length.
2. Instead of passing the map itself to the validation function, we could update the function signature to just accept the length of the map.
    - **Drawback w/ this approach**: Need to update validation-gen framework to pass in len(obj) for this (non-trivial change but definitely do-able).
3. We could do some `unsafe` pointer hack and spoof the map length while not actually creating a large map.


Going to keep this as is for now, lmk if there is an option you prefer from this or if you had any alternative ideas here for how to test this case.  Thanks!

Comment (aaron-prindle) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: Actually I think it is likely ok to use the `unsafe` method and spoof the map to be large.  I see `unsafe` used in other tests in k/k so I don't think it is an anti-pattern here.  I've updated the tests cases as suggested now and have opted to use `unsafe` to spoof the large map to support the test cases.

Comment (jpbetz) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: The fact that we can't produce a test for the max says something about the max.  Should we put our own artificial limit on the highest value maxProperties can be set to?

Comment (aaron-prindle) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: +1, that definitely makes sense of sense. If we can't practically test it, it's likely too high for any real use case.

I suggest setting the artificial limit to **100,000**.  I have updated the PR to enforce this 100,000 limit in the generator and cleaned up the tests accordingly.

**Rationale for 100,000**:
Given that the default etcd request size limit is 1.5 MB, even if we assume very small properties (eg: 10-15 bytes for key/value overhead in JSON - "k":"v"), a map could not exceed roughly 100,000 to 150,000 properties before the entire object becomes too large to store in etcd anyway. Can definitely tweak the number here if there are alternative thoughts on the max limit here (should be base 2, should consider different size limits, etc.)

I have removed the `unsafe` pointer hack from the tests now ( as we no longer need to spoof `math.MaxInt` to cover the edge case) and now test the new upper bound edge case of 100,000

```
	}, {
		name:       "100000 properties, max 100000",
		properties: 100000,
		max:        100000,
	}, {
		name:       "100001 properties, max 100000",
		properties: 100001,
		max:        100000,
		wantErrs: field.ErrorList{
			field.TooMany(field.NewPath("fldpath"), 100001, 100000).WithOrigin("maxProperties"),
		},

	}, {
```


Updated the Docs as well related to this:
```go
Docs:        "This map must have no more than X properties (where X <= 100000).",
```

Comment (jpbetz) [reply staging/src/k8s.io/apimachinery/pkg/api/validate/limits_test.go:213]: SG. I doubt anyone will complain, and if they think they need more than 100,000, we should probably be having a conversation about it.

**Comment by lalitc375** [comment]
/lgtm

Had a question about limiting the scope to string keyed maps.  This is done to align with the OpenAPI maxProperties. Most types in k/k have string as the key; it should be fine.
